flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/ZeroCopyNettyMessageDecoder.java

```diff
@@ -0,0 +1,280 @@
+
+    @Override
+    public void channelActive(ChannelHandlerContext ctx) throws Exception {
+        super.channelActive(ctx);
+
+        frameHeaderBuffer = ctx.alloc().directBuffer(NettyMessage.FRAME_HEADER_LENGTH);
+        messageHeaderBuffer = ctx.alloc().directBuffer(INITIAL_MESSAGE_HEADER_BUFFER_LENGTH);
+
+        decodeStep = DecodeStep.DECODING_FRAME_HEADER;
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        super.channelInactive(ctx);
+
+        if (intermediateState.messageHeaderParseResult != null) {
+            Buffer buffer = intermediateState.messageHeaderParseResult.getTargetDataBuffer();
+
+            if (buffer != null) {
+                buffer.recycleBuffer();
+            }
+        }
+
+        clearState();
+
+        frameHeaderBuffer.release();
+        messageHeaderBuffer.release();
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        if (!(msg instanceof ByteBuf)) {
+            ctx.fireChannelRead(msg);
+            return;
+        }
+
+        ByteBuf data = (ByteBuf) msg;
+
+        try {
+            while (data.readableBytes() > 0) {
+                if (decodeStep == DecodeStep.DECODING_FRAME_HEADER) {
+                    ByteBuf toDecode = cumulateBufferIfNeeded(frameHeaderBuffer, data, FRAME_HEADER_LENGTH);
+
+                    if (toDecode != null) {
+                        decodeFrameHeader(toDecode);
+
+                        decodedBytesOfCurrentStep = 0;
+                        decodeStep = DecodeStep.DECODING_MESSAGE_HEADER;
+                    }
+                }
+
+                if (decodeStep == DecodeStep.DECODING_MESSAGE_HEADER) {
+                    ByteBuf toDecoder = cumulateBufferIfNeeded(messageHeaderBuffer, data, intermediateState.messageHeaderLength);
+
+                    if (toDecoder != null) {
+                        intermediateState.messageHeaderParseResult = messageParser.parseMessageHeader(intermediateState.msgId, toDecoder);
+
+                        if (intermediateState.messageHeaderParseResult.getDataBufferAction() ==
+                                NettyMessageParser.DataBufferAction.NO_DATA_BUFFER) {
+                            ctx.fireChannelRead(intermediateState.messageHeaderParseResult.getParsedMessage());
+                            clearState();
+                        } else {
+                            decodedBytesOfCurrentStep = 0;
+                            decodeStep = DecodeStep.DECODING_BUFFER;
+                        }
+                    }
+                }
+
+                if (decodeStep == DecodeStep.DECODING_BUFFER) {
+                    readOrDiscardBufferResponse(data);
+
+                    if (decodedBytesOfCurrentStep == intermediateState.messageHeaderParseResult.getDataBufferSize()) {
+                        ctx.fireChannelRead(intermediateState.messageHeaderParseResult.getParsedMessage());
```

Comment:
If we can also cancel the partition directly while requesting the null buffer from `RemoteInputChannel` as concerned in https://github.com/apache/flink/pull/7368#pullrequestreview-354930094, then we do not need to fire it to the upper handler.

The `NetworkClientHandler` interface already provides the cancel request method. If the decoder covers this component which can be also used for getting input channel for requesting buffer in allocator, then it is feasible to do that.